x/campaign/keeper/msg_server_unredeem_vouchers.go

```diff
@@ -0,0 +1,58 @@
+package keeper
+
+import (
+	"context"
+
+	sdk "github.com/cosmos/cosmos-sdk/types"
+	sdkerrors "github.com/cosmos/cosmos-sdk/types/errors"
+	spnerrors "github.com/tendermint/spn/pkg/errors"
+	"github.com/tendermint/spn/x/campaign/types"
+)
+
+func (k msgServer) UnredeemVouchers(goCtx context.Context, msg *types.MsgUnredeemVouchers) (*types.MsgUnredeemVouchersResponse, error) {
+	ctx := sdk.UnwrapSDKContext(goCtx)
+
+	_, found := k.GetCampaign(ctx, msg.CampaignID)
+	if !found {
+		return nil, sdkerrors.Wrapf(types.ErrCampaignNotFound, "%d", msg.CampaignID)
+	}
+
+	account, found := k.GetMainnetAccount(ctx, msg.CampaignID, msg.Sender)
+	if !found {
+		return nil, sdkerrors.Wrap(types.ErrAccountNotFound, msg.Sender)
+	}
+
+	// Update the shares of the account
+	var err error
+	account.Shares, err = types.DecreaseShares(account.Shares, msg.Shares)
+	if err != nil {
+		return nil, sdkerrors.Wrap(types.ErrSharesDecrease, err.Error())
+	}
+
+	// If the account no longer has shares, it can be removed from the store
+	if types.IsEqualShares(account.Shares, types.EmptyShares()) {
+		k.RemoveMainnetAccount(ctx, msg.CampaignID, msg.Sender)
+	} else {
+		k.SetMainnetAccount(ctx, account)
+	}
```

Comment:
same comment as [here](https://github.com/tendermint/spn/pull/326/files#r720307032). we should save the account only if the account were already minted

Reply:
As mentioned, the context is only saved if the message is successful. In this case here, I think it's more visual to make the verification just after having decreased the shares